CHANGELOG.md

```diff
@@ -1,3 +1,10 @@
+## 4.0.1
```

Comment:
Seems like there are enough code changes here (esp. around thread safety and potential to leak keys in logs) to justify a minor (4.1.0)

Reply:
There are a lot of changes to the implementation details, but zero external features added (minor) and zero breaking changes (major); it exclusively fixes things that should have worked in the first place so I consider it to be a patch.